hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/translators/BlockStreamTransactionTranslator.java

```diff
@@ -0,0 +1,290 @@
+                blockItems.add(transactionOutput);
+            }
+            blockItems.add(stateChanges);
+
+            return blockItems;
+        }
+
+        /**
+         * The input block items should be exactly the block items produced for/by processing a
+         * single transaction, with the following expected order:
+         * <ol>
+         *     <li>Index 0: BlockItem.Transaction</li>
+         *     <li>Index 1: BlockItem.TransactionResult</li>
+         *     <li>Index 2: BlockItem.TransactionOutput (if applicable)</li>
+         *     <li>Index 2 or 3: BlockItem.StateChanges (dependent on the presence of TransactionOutput)</li>
+         * </ol>
+         *
+         * @param items The block items representing a single transaction
+         * @return A logical transaction wrapper for the block items
+         */
+        public static BlockTransaction asBlockTransaction(@NonNull List<BlockItem> items) {
+            final var txn = items.get(0);
+            final var result = items.get(1);
+            final var maybeOutput = items.get(2);
+
+            if (items.size() > 3) {
+                return new BlockTransaction(txn, result, maybeOutput, items.get(3));
+            } else {
+                return new BlockTransaction(txn, result, null, maybeOutput);
+            }
+        }
+    }
+
+    /**
+     * Translates a {@link BlockTransaction} into a {@link TransactionRecord}.
+     *
+     * @param transaction A wrapper for block items representing a single txnInput
+     * @return the translated txnInput record
+     */
+    @Override
+    public TransactionRecord translate(@NonNull final BlockTransaction transaction) {
+        validateBlockItems(transaction.asItems());
+
+        final var recordBuilder = TransactionRecord.newBuilder();
+        final var receiptBuilder = TransactionReceipt.newBuilder();
+
+        BlockItem txnBlockItem = transaction.txnInput();
+        if (txnBlockItem.hasTransaction()) {
+            parseTransaction(txnBlockItem, recordBuilder);
+        }
+
+        txnBlockItem = transaction.transactionResult();
+        if (txnBlockItem.hasTransactionResult()) {
+            parseTransactionResult(txnBlockItem, recordBuilder, receiptBuilder);
+        }
+
+        txnBlockItem = transaction.transactionOutput();
+        if (txnBlockItem != null && txnBlockItem.hasTransactionOutput()) {
+            parseTransactionOutput(txnBlockItem, recordBuilder, receiptBuilder);
+        }
+
+        txnBlockItem = transaction.stateChanges();
+        // TODO: how do we generically extract state changes for the txnInput record?
+        if (txnBlockItem.hasStateChanges()) {}
+
+        return recordBuilder.build();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<TransactionRecord> translateAll(final List<BlockTransaction> transactions) {
+        throw new NotImplementedException();
+    }
+
+    private void validateBlockItems(final List<BlockItem> blockItems) {
```

Comment:
I don't get why this doesn't just take a `BlockTransaction` instead of turning it into a `List` first.  If a `BlockTranslation` already then you don't have to worry about index numbers here, information which seems properly encapsulated in `BlockTransaction.asBlockTransaction` factory method already.

(Only caller (at this time) already has the `BlockTransaction`.  If the issue is that maybe a different caller might someday only have the list of things then that argues for this `validateBlockItems` method being a member of the record `BlockTransaction` itself and called inside of `asBlockTransaction` first thing.)

Reply:
The reason I wrote it this way is because this method came before `asBlockTransaction`, and I didn't think to change it :) good call, will do

Reply:
Done